driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java

```diff
@@ -0,0 +1,173 @@
+ * The first potentially missing value within each interval is its lower bound, other values are computed by adding the step
+ * multiple times, until the result is out of the interval. Each time the step is added, the result is a potentially missing value for
+ * which a new document must be added if the sequence of documents that is being densified does not have a document
+ * with equal value of the field.
+ *
+ * @mongodb.server.release 5.1
+ * @since 4.7
+ */
+@Evolving
+public interface DensifyRange extends Bson {
+    /**
+     * Returns a {@link DensifyRange} that represents an interval with the smallest
+     * BSON {@link BsonType#INT32 32-bit integer} / {@link BsonType#INT64 64-bit integer} / {@link BsonType#DOUBLE Double} /
+     * {@link BsonType#DECIMAL128 Decimal128} value of the {@linkplain DensifySpecification#densifyField(String) field}
+     * in the sequence of documents being its lower bound, and the largest value being the upper bound.
+     *
+     * @param step The positive step.
+     * @return The requested {@link DensifyRange}.
+     */
+    static NumberDensifyRange numberRangeFullWithStep(final Number step) {
+        return new DensifyConstructibleBson(new Document("bounds", "full")
+                .append("step", notNull("step", step)));
+    }
+
+    /**
+     * Returns a {@link DensifyRange} that represents an interval with the smallest
+     * BSON {@link BsonType#INT32 32-bit integer} / {@link BsonType#INT64 64-bit integer} / {@link BsonType#DOUBLE Double} /
+     * {@link BsonType#DECIMAL128 Decimal128} value of the {@linkplain DensifySpecification#densifyField(String) field}
+     * in the {@linkplain DensifySpecification#partitionByFields(Iterable) partition} of documents being its lower bound,
+     * and the largest value being the upper bound.
+     *
+     * @param step The positive step.
+     * @return The requested {@link DensifyRange}.
+     */
+    static PartitionByFieldNumberDensifyRange numberRangePartitionWithStep(final Number step) {
+        return new DensifyConstructibleBson(new Document("bounds", "partition")
+                .append("step", notNull("step", step)));
+    }
+
+    /**
+     * Returns a {@link DensifyRange} that represents a single interval [l; u).
+     *
+     * @param l The lower bound.
+     * @param u The upper bound.
+     * @param step The positive step.
+     * @return The requested {@link DensifyRange}.
+     */
+    static NumberDensifyRange numberRangeWithStep(final Number l, final Number u, final Number step) {
+        notNull("l", l);
+        notNull("u", u);
+        if (l instanceof Comparable && l.getClass().isAssignableFrom(u.getClass())) {
+            @SuppressWarnings("unchecked")
+            Comparable<Number> comparableL = (Comparable<Number>) l;
+            // a half open interval with equal bounds is an empty interval, and the server allows specifying it
```

Comment:
And we don't prevent it, we allow it by including `=` in the `<=` check. I left a comment because I found this surprising, and wanted to explain why the check allows specifying an empty interval.

Reply:
oh, i see.  But do we need the check at all?  If the server reports an error when `l > u` then we don't need to do so in the driver.  This is consistent with general driver practice of allowing the server to enforce constraints.

Reply:
Done.